### ISSUE TITLE
Subsonic connection fails using DSub

### ISSUE THREAD

**Issue by Biont** [owncloud/music#720]
I believe that I setup everything according to the docs and testing the connection within the app worked at first. However, everything after that initial test fails and I get a load of errors like this:

```
{"reqId":"S2MxCCd65H0clUZ2sdbG","level":2,"time":"2019-10-13T11:57:45+00:00","remoteAddr":"xxxxxxxxx","user":"--","app":"core","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getUser.view?u=xxxxxx&p=enc:xxxxxxxxxxxxxxxxxxxxx&v=1.2.0&c=DSub&username=xxxx","message":"Login failed: 'xxxxxxxx' (Remote IP: xxxxxxxxxxx')","userAgent":"DSub","version":"17.0.0.9"}
{"reqId":"S2MxCCd65H0clUZ2sdbG","level":1,"time":"2019-10-13T11:57:45+00:00","remoteAddr":"xxxxxxxxx","user":"--","app":"core","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getUser.view?u=xxxxxx&p=enc:xxxxxxxxxxxxxxxxxxxxx&v=1.2.0&c=DSub&username=xxxxxxx","message":"Bruteforce attempt from \"xxxxxxxxxx\" detected for action \"login\".","userAgent":"DSub","version":"17.0.0.9"}
{"reqId":"S2MxCCd65H0clUZ2sdbG","level":2,"time":"2019-10-13T11:57:45+00:00","remoteAddr":"xxxxxxxxx","user":"--","app":"music","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getUser.view?u=xxxxxxx&p=enc:xxxxxxxxxxxxxxxxxxxxx&v=1.2.0&c=DSub&username=xxxxxx","message":"Request getUser not supported","userAgent":"DSub","version":"17.0.0.9"}
```
 I will happily provide any additional information you might need. Thanks for implementing this!

**Comment by paulijar** [comment]
Thanks for the report.

You seem to be using DSub v17.0.0.9 while I have been testing v13.0.5.2. I don't know if that's relevant here, though.

One thing is, that at least my version of DSub has an option "Browse By Tags" in the server settings. This must be unticked, because otherwise DSub uses such API methods which are not implemented on the Music app.

The warning "Request getUser not supported" is something I'm getting too, and I probably should add support for this Subsonic API method. But for me, this doesn't seem to cause any issue on the DSub UI.

Have you already tried other Subsonic clients? I've been mostly using [Subsonic Music Streamer](https://play.google.com/store/apps/details?id=net.sourceforge.subsonic.androidapp&hl=en) as the reference client while implementing the API.

**Comment by paulijar** [comment]
BTW, be careful while pasting log entries from the Subsonic calls. You had obfuscated some of the fields, but the argument `p=enc:3069...` was showing your generated API password in HEX format which is trivial to decode. I now obfuscated that for you. But you still should drop that password and generate a new one.

**Comment by Biont** [comment]
Thanks for the hint.
The "browse by tags" option is unticked.


I can try the other app later tonight, but I was hoping to get DSub working since it's the only one to support chromecasts as far as I know.

I post if it works later.

**Comment by paulijar** [comment]
About your edit: Okay, I didn't understand that even the connection test stopped working, and those brute force attempt messages might well be significant here. Note that they come from 'core' and not from 'music'. It could be that DSub is now repeating some request so frequently after a failure, that the core thinks this is a brute force attempt and blocks the traffic. During the development of the Subsonic feature I sometimes saw such warnings when I was still missing many needed REST endpoints, but I haven't seen them lately.

I now implemented the getUser endpoint and pushed it to master. You could give it a try and see if it would happen to help here.

**Comment by Biont** [comment]
@paulijar Thanks for the quick response. I checked out the current master on my server and it seems to have helped quite a bit. I can browse the library and even get playback. 

But there are still network errors every now (-> when it will retry 4 times and my logs are getting spammed) and then and the whole thing feels a bit sluggish and unstable.

Here are some logs:

First, there's a PHP error that I get quite often 
`{"reqId":"onLTr3Xl7N5hrZK48mhm","level":3,"time":"2019-10-13T19:13:25+00:00","remoteAddr":"","user":"--","app":"PHP","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getMusicDirectory.view?u=xxxxxxx&p=enc:XXXXXXXXXXXXXXX&v=1.2.0&c=DSub&id=album-1377","message":"Only variables should be passed by reference at \/usr\/share\/webapps\/nextcloud\/apps\/music\/controller\/subsoniccontroller.php#687","userAgent":"DSub","version":"17.0.0.9"}
`
...and then there's still the occasional bruteforce warning and failed logins:

`{"reqId":"dA20AU8joEx0OoZLRIsq","level":1,"time":"2019-10-13T19:13:07+00:00","remoteAddr":"XXXXXXXXX","user":"--","app":"core","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getCoverArt.view?u=XXXXXXX&p=enc:31366e7365623932766a7263&v=1.2.0&c=DSub&id=1377","message":"Bruteforce attempt from \"92.117.111.31\" detected for action \"login\".","userAgent":"DSub","version":"17.0.0.9"}
{"reqId":"9dWk64FX3iNiwflbRfz0","level":2,"time":"2019-10-13T19:13:09+00:00","remoteAddr":"92.117.111.31","user":"--","app":"core","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getCoverArt.view?u=XXXXXXX&p=enc:XXXXXXXXXXXX&v=1.2.0&c=DSub&id=1199","message":"Login failed: 'XXXXXX' (Remote IP: 'XXXXXXXXXXX')","userAgent":"DSub","version":"17.0.0.9"}
{"reqId":"9dWk64FX3iNiwflbRfz0","level":1,"time":"2019-10-13T19:13:09+00:00","remoteAddr":"92.117.111.31","user":"--","app":"core","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getCoverArt.view?u=XXXXXXXX&p=enc:31366e7365623932766a7263&v=1.2.0&c=DSub&id=1199","message":"Bruteforce attempt from \"XXXXXXXXX\" detected for action \"login\".","userAgent":"DSub","version":"17.0.0.9"}
{"reqId":"onLTr3Xl7N5hrZK48mhm","level":2,"time":"2019-10-13T19:13:25+00:00","remoteAddr":"92.117.111.31","user":"--","app":"core","method":"GET","url":"\/index.php\/apps\/music\/subsonic\/rest\/getMusicDirectory.view?u=XXXXXX&p=enc:XXXXXXXXXXXXXXX&v=1.2.0&c=DSub&id=album-1377","message":"Login failed: 'XXXXXXX' (Remote IP: 'XXXXXXXX')","userAgent":"DSub","version":"17.0.0.9"}
`


I also gave the Subsonic Music Streamer App a try and that seems to work more or less flawlessly.

**Comment by paulijar** [comment]
The message `Only variables should be passed by reference` appeared because I was calling the PHP built-in `end()` function a bit incorrectly. This should have caused no other harm than the annoying message. Anyway, I now fixed this.

It's still a mystery, why the core sometimes thinks that the random Subsonic API calls are login attempts. 

Which cloud version are you using?

**Comment by Biont** [comment]
Nextcloud 17.0.0 @paulijar

**Comment by nachoparker** [comment]
NC 16.0.5 here, (NextCloudPi)

Just tried 0.11 with the Subsonic API and it is working for me with Dsub. Thanks!

There are other two Android clients on F-Droid: Ultrasonic and Audinaut

I also tried Ultrasonic, but it would not connect, probably uses a different version of the API. DSub seems to work fine and even though it has not been updated in 4 years it seems quite feature complete.

Audinaut also doesn't connect. Both Ultrasonic and Audinaut seem forked from DSub or at least from each other. At least the settings menu is identical.

I have been using Power Ampache until now, but it is nice that I can now browse by folders on my phone, and Dsub doesn't struggle with my big collection, as PA used to do before, even though they have improved this.

Very happy overall, even though the choice for Android FOSS users has increased from Power Ampache to PA + DSub, it is a shame that the other two alternatives don't work. 

I have not tested them, since they don't work with Nextcloud Music but they have been maintained actively until more recently so I assume it would be nice for FOSS users if they were also supported, to have more choice.

All in all this is a great improvement. Thanks

I think it's time to write a follow up to 

https://ownyourbits.com/2018/08/28/stream-you-music-with-nextcloud-music-and-power-ampache/

Would you have some comments about how subsonic API compares to good old Ampache API so I can include it?

Thanks for this. Big step in usability for me.

**Comment by Biont** [comment]
For the record, I have temporarily disabled brute force protection and still get the "Login failed:" errors. So it seems that the logs from core are an effect of the failed login attempts, not their cause.

**Comment by Biont** [comment]
It looks like DSub sends an `Authorization` header with each(?) request. See [here](https://github.com/daneren2005/Subsonic/blob/166d395e3e5de223527fec96d8b17dd42ed9a991/app/src/main/java/github/daneren2005/dsub/service/RESTMusicService.java#L1918)

I have verified by adding a ` $this->logger->warning('DSUBREQUEST '. json_encode(\OC::$server->getRequest()->getHeader('Authorization')), ['app'=>'core']);
` right after the "Login failed: " line in `/lib/private/User/Session.php` (line 444 in my version). And yup, it contains a `Basic XXXXXXXXXXXXXXX` payload.

I have not checked what is actually sent there, but my guess is that it's simply the app password that we configured in the settings of the music app as some sort of safety net. Nextcloud will attempt to log your user in, fail, and trigger the bruteforce attempt. 

Then your SubSonicMiddleware kicks in and handles the request normally using the credentials from the GET params.

I lack experience with Nextcloud to know how to move on with this, but I hope this information will be helpful to you and that you are able to figure something out

**Comment by paulijar** [comment]
I don't have time to comment on all the recent posts for now, but I'll come back to them later this week. But briefly about Ultrasonic and Audinaut: I tested both of these apps during the development but I couldn't make Ultrasonic work at all, even with the official Subsonic demo server. It didn't just refuse to connect, it also didn't let me open the app any more after the first unsuccessful connection attempt! I think the app is just plain broken, or then it just does not work on my Android version.

Audinaut's problem, on the other hand, is that it only supports the token-based authentication introduced in Subsonic API v13. I think this is an oversight in the client and should be fixed there: even the real Subsonic server may be unable to use the token-based authentication in case it is set up with LDAP. There's even dedicated error code for that situation in the API spec, and we are returning that code in case the client attempts to use the token authentication. I really wouldn't want to support the token authentication for reasons described in #718.

**Comment by nachoparker** [comment]
Yeah makes sense. I did not know if there's stuff missing on the API or what it was on each case. Sadly not a ton of options for Android but at least we have dsub

**Comment by paulijar** [comment]
@Biont 
> Nextcloud 17.0.0

Right, that was right there visible in your logs. I initially misinterpreted the log and thought that `"version":"17.0.0.9"` in the log referred to the client, but that wasn't the case. Are you using the most recent version of DSub? I noticed that F-Droid is offering a very old version 5.0.3 while Play Store has the most recent version 5.4.4.

**Comment by Biont** [comment]
@paulijar I am using the play store version 5.4.4

**Comment by jakobroehrl** [comment]
Hi, same problem here. 
With the 5.4.4 (play store) I can't connect. 
But with the older f-droid version it's working

**Comment by Biont** [comment]
Comparing https://github.com/daneren2005/Subsonic/blob/166d395e3e5de223527fec96d8b17dd42ed9a991/app/src/main/java/github/daneren2005/dsub/service/RESTMusicService.java#L1918

with the same file from version 5.0.3:
https://github.com/daneren2005/Subsonic/blob/5.0.3/app/src/main/java/github/daneren2005/dsub/service/RESTMusicService.java

....shows that the basic auth header was not present and has been added by this commit:
https://github.com/daneren2005/Subsonic/commit/ed010a2188f58e0581a95506cf7eecc1cac35d9c

I guess unless there is a way to tell NC to ignore the Basic Auth method for the Subsonic REST routes, we need to convince the DSub dev(s) to make this feature optional.

**Comment by paulijar** [comment]
This has been requested on DSub already a year ago: https://github.com/daneren2005/Subsonic/issues/870

It hasn't sprung any kind of discussion, though.

**Comment by paulijar** [comment]
So the conclusion is that DSub installed from F-Droid works fine, but the version bought from Play Store does not work. It's likely, that we cannot do anything on Music app to fix the compatibility. The best chance to get this fixed would probably be, if someone affected would contribute a PR for https://github.com/daneren2005/Subsonic/issues/870.

**Comment by Biont** [comment]
Man I just want to chromecast my Nextcloud Music :'(
I mean it works if I keep brute-force detection disabled, but that's not a tradeoff I'm willing to make in the long run.

**Comment by paulijar** [comment]
Is using the DSub 5.0.3 an alternative for you, or is that version missing the chromecast support?

As a workaround on DSub 5.4.4, you could set your Nextcloud password to be the same as the generated Subsonic API password (*). To minimize the security impact of this, you should create a new Nextcloud user just for this purpose. Then share your music collection directory in read-only mode with this new user. Quite clumsy, I know, but seems to work as I tested this with DSub 5.4.4.

(*) Or vice versa. You can't change the Subsonic API passwords through the UI, but it's possible by directly modifying the contents of the database table `oc_music_ampache_users`. You should calculate the sha256 hash of your new password and store it in the column `hash`.

**Comment by jakobroehrl** [comment]
Getting Nextcloud and chromecast working is my goal, too.
With the password hack I can use the 5.4.4 and can connect to chromecast.
But the connection is not very stabile, I get loading errors. :-(

**Comment by paulijar** [comment]
@jakobroehrl Sorry, there's no support for Sonos.

**Comment by JoshuaPettus** [comment]
FTR, since I didn't see it mentioned.   I gave 5.0.3 on FDroid a try, it seems that version could have had chromecast, but I read on one of the issue threads for dsub that the FDroid version was compiled without with the necessary google libraries that the PlayStore version.  So the button is there, but the feature is non functional.  At least that has been my experience with it.  Shame though, I was hoping it would work :)  Also shame I cant downgrade to 5.0.3 on the play store or I'd consider paying the 4 bucks.  

On a personal note.  I'll still keep the FDroid DSub app for now, as I find it more reliable than subsonic's own android app.  Sometimes subsonic will stop in the middle of a song and skip, or worse, restart the song which is real annoying with FLAC tracks which can't be scrolled through.  (Something I've noticed is true with VLC too so it's not just you).

**Comment by paulijar** [comment]
> Audinaut's problem, on the other hand, is that it only supports the token-based authentication

Update: Audinaut v0.5.0 available [here](https://github.com/nvllsvm/Audinaut/releases) now supports disabling the token/salt authentication and using the plain password authentication instead. With this setting, the application can communicate with the ownCloud Music back-end. I didn't test it extensively, but at least it showed my library and played some songs.

**Comment by JoshuaPettus** [comment]
Oooh  I'll have to give it a try! Thanks!

**Comment by merlin-tc** [comment]
> Update: Audinaut v0.5.0 available [here](https://github.com/nvllsvm/Audinaut/releases) now supports disabling the token/salt authentication and using the plain password authentication instead. With this setting, the application can communicate with the ownCloud Music back-end. I didn't test it extensively, but at least it showed my library and played some songs.

@paulijar 
Just tried Audinaut (have been using dsub before) and it works like a charm.
Thanks for the suggestion and of course for your work on the music nextcloud plugin. I am using it daily!

**Comment by paulijar** [comment]
Just for completeness, and because UltraSonic has been mentioned in this thread before, it has been supported by the Music app since v0.14.1.

**Comment by paulijar** [comment]
Here's one more piece of information related to DSub: All the versions are available for download from https://github.com/daneren2005/Subsonic/releases. I just tested that version 5.3.2 is the latest one still working with the Music app without issues. That is, version 5.3.3 introduced the breaking change of adding the basic authorization header. 

(Okay, I just realized that there would have been an easier way to find this out than testing, because the culprit commit was linked in [a post above](https://github.com/owncloud/music/issues/720#issuecomment-543701750), and github shows, which versions include that commit. But I did it the hard way 😆.)
